lib/compiler/src/sys_core_alias.erl

```diff
@@ -43,13 +43,29 @@
 -include("core_parse.hrl").
 
 -define(NOTSET, 0).
+-define(HAS_SUBS(Sub), Sub#sub.t =/= none).
 
 -record(sub, {p=#{} :: #{term() => ?NOTSET | atom()},              %% Found pattern substitutions
               v=cerl_sets:new() :: cerl_sets:set(cerl:var_name()), %% Variables used by patterns
-              t=undefined :: term()}).                             %% Temporary information from pre to post
+              t=none :: temp()}).                                  %% Temporary information from pre to post
 
 -type sub() :: #sub{}.
 
+%% temp keeps the stack of substitutions.
+%%
+%% 1. none is the root
+%% 2. temp stores the previous temp value - this is used when
+%%    we update the existing sub instead of replacing it
+%% 3. sub is the previous substitution - this is used when
+%%    we replace the current sub
+%% 4. clause is used when there are clauses with patterns
+%%
+-type temp() ::
+        none
+        | {temp, temp()}
+        | {sub, sub()}
+        | {clause, [term()], [term()], cerl_sets:new(), temp()}.
```

Comment:
Thanks for the PR!

`dialyzer` is unhappy with `cerl_sets:new()`. Travis caught this but the notifications are broken for some reason, sorry about that. :(

```suggestion
        | {clause, [term()], [term()], cerl_sets:set(cerl:var_name()), temp()}.
```